From 97d0da556e5ae568f9fc13763ab2b6c5cf1c4a8a Mon Sep 17 00:00:00 2001
From: Mahesh R Vaidya <mahesh.r.vaidya@intel.com>
Date: Thu, 8 May 2025 02:40:43 +0530
Subject: [PATCH] Config-3 concurrent 3x2.5G Device tree changes

        1. Add 3 MRPHY gmii-to-sgmii adapters nodes for gmac0(mdio),
	   gmac1(mdio) and gmac2(i2c-sfp) to support concurrent(3x2.5G).
        2. Add support for gmii mac mode and 2500base-x phy mode to all
	   3 gmac nodes requried for concurrent(3x2.5G) TSN CFG-3
	3. Add altera mdio controller support to GMAC0 and GMAC1 PHYs
	   and remove the stmmac mdio support on these PHYs.
	4. Add sfp node support to GMAC2.
	5. Add altera i2c node device support required by sfp for GMAC2
	   management.

Upstream-Status: Pending
Signed-off-by: Mahesh R Vaidya <mahesh.vaidya@altera.com>
---
 arch/arm64/boot/dts/intel/Makefile            |   1 +
 .../boot/dts/intel/socfpga_agilex5_socdk.dts  |  12 ---
 .../intel/socfpga_agilex5_socdk_tsn_cfg3.dts  | 101 ++++++++++++++++++
 3 files changed, 102 insertions(+), 12 deletions(-)
 create mode 100644 arch/arm64/boot/dts/intel/socfpga_agilex5_socdk_tsn_cfg3.dts

diff --git a/arch/arm64/boot/dts/intel/Makefile b/arch/arm64/boot/dts/intel/Makefile
index 1f62d0e4f3db..d7b40080f702 100644
--- a/arch/arm64/boot/dts/intel/Makefile
+++ b/arch/arm64/boot/dts/intel/Makefile
@@ -15,6 +15,7 @@ dtb-$(CONFIG_ARCH_INTEL_SOCFPGA) += socfpga_agilex_n6000.dtb \
 				socfpga_agilex5_socdk_swvp_b0.dtb \
 				socfpga_agilex5_socdk_tsn_cfg2.dtb \
 				socfpga_agilex5_socdk_tsn_cfg2_b0.dtb \
+				socfpga_agilex5_socdk_tsn_cfg3.dtb \
 				socfpga_agilex5_socdk_modular.dtb \
 				socfpga_agilex7m_socdk.dtb \
 				socfpga_agilex7f_socdk_pcie_root_port.dtb \
diff --git a/arch/arm64/boot/dts/intel/socfpga_agilex5_socdk.dts b/arch/arm64/boot/dts/intel/socfpga_agilex5_socdk.dts
index 67bdf5142422..4bbbc3efdb73 100644
--- a/arch/arm64/boot/dts/intel/socfpga_agilex5_socdk.dts
+++ b/arch/arm64/boot/dts/intel/socfpga_agilex5_socdk.dts
@@ -51,7 +51,6 @@ &gpio1 {
 &gmac0 {
 	status = "disabled";
 	phy-mode = "rgmii";
-	phy-handle = <&emac0_phy0>;
 
 	max-frame-size = <9000>;
 
@@ -59,9 +58,6 @@ mdio0 {
 		#address-cells = <1>;
 		#size-cells = <0>;
 		compatible = "snps,dwmac-mdio";
-		emac0_phy0: ethernet-phy@0 {
-			reg = <0>;
-		};
 	};
 };
 
@@ -69,7 +65,6 @@ &gmac1 {
 	status = "disabled";
 	phy-mode = "rgmii";
 	mac-mode = "gmii";
-	phy-handle = <&emac1_phy0>;
 
 	max-frame-size = <9000>;
 
@@ -77,16 +72,12 @@ mdio0 {
 		#address-cells = <1>;
 		#size-cells = <0>;
 		compatible = "snps,dwmac-mdio";
-		emac1_phy0: ethernet-phy@0 {
-			reg = <0>;
-		};
 	};
 };
 
 &gmac2 {
 	status = "okay";
 	phy-mode = "rgmii";
-	phy-handle = <&emac2_phy0>;
 
 	max-frame-size = <9000>;
 
@@ -94,9 +85,6 @@ mdio0 {
 		#address-cells = <1>;
 		#size-cells = <0>;
 		compatible = "snps,dwmac-mdio";
-		emac2_phy0: ethernet-phy@0 {
-			reg = <0>;
-		};
 	};
 };
 
diff --git a/arch/arm64/boot/dts/intel/socfpga_agilex5_socdk_tsn_cfg3.dts b/arch/arm64/boot/dts/intel/socfpga_agilex5_socdk_tsn_cfg3.dts
new file mode 100644
index 000000000000..292c34d22beb
--- /dev/null
+++ b/arch/arm64/boot/dts/intel/socfpga_agilex5_socdk_tsn_cfg3.dts
@@ -0,0 +1,101 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * Copyright (c) 2024, Intel Corporation
+ */
+#include "socfpga_agilex5_socdk.dts"
+
+/ {
+	gmii_to_sgmii_converter0: phy@30020100 {
+		compatible = "altr,gmii-to-sgmii-2.0";
+		reg = <0x0 0x30020100 0x0 0x80>;
+		reg-names = "mrphy";
+	};
+
+	gmii_to_sgmii_converter1: phy@30020180 {
+		compatible = "altr,gmii-to-sgmii-2.0";
+		reg = <0x0 0x30020180 0x0 0x80>;
+		reg-names = "mrphy";
+	};
+
+	gmii_to_sgmii_converter2: phy@30020200 {
+		compatible = "altr,gmii-to-sgmii-2.0";
+		reg = <0x0 0x30020200 0x0 0x80>;
+		reg-names = "mrphy";
+	};
+
+	mdio: mdio@0x30020500 {
+		compatible = "altr, altr-mdio";
+		reg = <0x0 0x30020500 0x0 0x100>;
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		phy_0: ethernet-phy@0 {
+			reg = <0>;
+			compatible = "ethernet-phy-ieee802.3-c45";
+		};
+
+		phy_1: ethernet-phy@1 {
+			reg = <1>;
+			compatible = "ethernet-phy-ieee802.3-c45";
+		};
+	};
+
+	i2c5: i2c@30020600 {
+		compatible = "altr,softip-i2c-v1.0";
+		reg = <0x0 0x30020600 0x0 0x100>;
+		#address-cells = <1>;
+		#size-cells = <0>;
+		interrupt-parent = <0x03>;
+		interrupts = <0x00 0x33 0x04>;
+		clocks = <&clkmgr AGILEX5_L4_SP_CLK>;
+		fifo-size = <0x100>;
+		clock-frequency = <400000>;
+	};
+};
+
+&gmac0 {
+	status = "okay";
+	phy-mode = "2500base-x";
+	mac-mode = "gmii";
+	phy-handle = <&phy_0>;
+	max-speed = <2500>;
+	max-frame-size = <9000>;
+	#address-cells = <1>;
+	#size-cells = <1>;
+	altr,gmii-to-sgmii-converter = <&gmii_to_sgmii_converter0>;
+};
+
+&gmac1 {
+	status = "okay";
+	phy-mode = "2500base-x";
+	mac-mode = "gmii";
+	phy-handle = <&phy_1>;
+	max-speed = <2500>;
+	max-frame-size = <9000>;
+	#address-cells = <1>;
+	#size-cells = <1>;
+	altr,gmii-to-sgmii-converter = <&gmii_to_sgmii_converter1>;
+};
+
+&gmac2 {
+	status = "okay";
+	phy-mode = "2500base-x";
+	mac-mode = "gmii";
+	max-speed = <2500>;
+	max-frame-size = <9000>;
+	#address-cells = <1>;
+	#size-cells = <1>;
+	altr,gmii-to-sgmii-converter = <&gmii_to_sgmii_converter2>;
+
+	sfp28_cage: sfp {
+		compatible = "sff,sfp";
+		i2c-bus = <&i2c5>;
+	};
+
+	fixed-link {
+		speed = <2500>;
+		full-duplex;
+	};
+};
+
+
-- 
2.34.1

